<h1 style="color:DarkTurquoise"> Exploratory Notebook </h1> 

Note: Please make sure to select the conda kernel. Otherwise, this won't work correctly (`statsmodels` is not 
installed in my base `/usr` root)

Import necessary package dependencies

In [1]:
import pandas as pd
import statsmodels
import os

In [2]:
import sys

sys.prefix

'/home/jjl20011/miniconda3'

In [3]:
data_path = "../../data/InsNova_data_2023_train.csv"

data = pd.read_csv(data_path)

initial dataframe exploration (variable exploration)

In [4]:
data.head()

id  veh_value  exposure veh_body  veh_age gender area  agecat engine_type  \
0   1       0.77  0.444504    SEDAN        4      M    D       3      petrol   
1   2       4.45  0.562183    STNWG        1      M    A       3      petrol   
2   3       4.90  0.465244    STNWG        1      F    A       3      petrol   
3   4       0.48  0.271039    PANVN        4      M    A       4      petrol   
4   5       0.85  0.141624    SEDAN        4      F    A       5      petrol   

   max_power  ...  marital_status e_bill time_of_week_driven  time_driven  \
0        147  ...               S      1             weekday   6pm - 12am   
1        158  ...               S      1             weekday   6am - 12pm   
2        159  ...               M      1             weekday   6pm - 12am   
3         80  ...               S      1             weekday   12pm - 6pm   
4        126  ...               S      0             weekday   6am - 12pm   

  trm_len credit_score  high_education_ind  clm  numclaims  claimcst0  
0       6   640.448137                 1.0    0          0        0.0  
1      12   683.749691                 0.0    0          0        0.0  
2       6   653.656117                 1.0    0          0        0.0  
3      12   642.574671                 0.0    0          0        0.0  
4       6   647.175035                 0.0    0          0        0.0  

[5 rows x 22 columns]

In [5]:
data.describe()

id     veh_value      exposure       veh_age        agecat  \
count  22619.000000  22619.000000  22619.000000  22619.000000  22619.000000   
mean   11310.000000      1.868948      0.433038      2.667492      3.481675   
std     6529.687205      1.278588      0.272899      1.070555      1.420252   
min        1.000000      0.000000      0.001754      1.000000      1.000000   
25%     5655.500000      1.070000      0.203696      2.000000      2.000000   
50%    11310.000000      1.570000      0.384313      3.000000      3.000000   
75%    16964.500000      2.260000      0.643316      4.000000      5.000000   
max    22619.000000     24.510000      0.999378      4.000000      6.000000   

          max_power  driving_history_score       e_bill       trm_len  \
count  22619.000000           22619.000000  22619.00000  22619.000000   
mean     152.447765              70.868341      0.63681     10.493833   
std       51.834156              19.083318      0.48093      2.601685   
min       60.000000               1.000000      0.00000      6.000000   
25%      115.000000              58.000000      0.00000      6.000000   
50%      144.000000              72.000000      1.00000     12.000000   
75%      180.000000              86.000000      1.00000     12.000000   
max      409.000000              99.000000      1.00000     12.000000   

       credit_score  high_education_ind           clm     numclaims  \
count  22619.000000        22619.000000  22619.000000  22619.000000   
mean     649.875077            0.117114      0.068173      0.073124   
std       10.515364            0.321563      0.252048      0.280142   
min      607.797435            0.000000      0.000000      0.000000   
25%      643.134748            0.000000      0.000000      0.000000   
50%      647.180197            0.000000      0.000000      0.000000   
75%      654.069205            0.000000      0.000000      0.000000   
max      790.360253            1.000000      1.000000      3.000000   

          claimcst0  
count  22619.000000  
mean     163.048084  
std     1271.955238  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max    57895.584560

**Insights**: 
Based on the summary, it looks as if the `claimscst0` variable, which indicates the claim cost. 
There are 22,619 samples in the data set. This should be sufficient to train a simple neural network 
if we wanted to. There is likely good actuarial research on building statistical claims models. These
could be good to look into as well. 

I think, the best place to start would be just to start with a general linear model (ie. multiple linear regression anova and analyze the goodness of the linear fit)

<h3 style="color:LightSalmon"> Bing AI Help </h3> 

Use the following to check the anova stuff

```
import statsmodels.api as sm
model = sm.OLS(y, X).fit()
type_iii_ss = model.ess
type_i_ss = model.ess - model.resid_ss

```

In [7]:
import statsmodels.api as sm

# the `endog` variable is the dependent (response) variable
# `exog` is the independent variable(s) - predictor variables
model = sm.OLS(endog=data['claimcst0'], exog=data['veh_value']).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              claimcst0   R-squared (uncentered):                   0.011
Model:                            OLS   Adj. R-squared (uncentered):              0.011
Method:                 Least Squares   F-statistic:                              259.7
Date:                Fri, 03 Nov 2023   Prob (F-statistic):                    4.21e-58
Time:                        13:58:23   Log-Likelihood:                     -1.9384e+05
No. Observations:               22619   AIC:                                  3.877e+05
Df Residuals:                   22618   BIC:                                  3.877e+05
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
veh_value     60.3350      3.744     16.115      0.000      52.996      67.673
==============================================================================
Omnibus:                    47232.484   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        250639553.131
Skew:                          17.887   Prob(JB):                         0.00
Kurtosis:                     517.454   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [19]:
import statsmodels.formula.api as smf

model2 = smf.ols(formula="clm ~ driving_history_score", data=data).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    clm   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.510
Date:                Fri, 03 Nov 2023   Prob (F-statistic):              0.219
Time:                        14:45:47   Log-Likelihood:                -921.64
No. Observations:               22619   AIC:                             1847.
Df Residuals:                   22617   BIC:                             1863.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 0.0605      0.006      9.390      0.000       0.048       0.073
driving_history_score     0.0001   8.78e-05      1.229      0.219   -6.42e-05       0.000
==============================================================================
Omnibus:                    15413.691   Durbin-Watson:                   2.014
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           133672.877
Skew:                           3.426   Prob(JB):                         0.00
Kurtosis:                      12.740   Cond. No.                         282.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
stats